bpf/cpu/pyperf.bpf.c

```diff
@@ -0,0 +1,439 @@
+
+#define LOG(fmt, ...)                                                                                                                                          \
+  ({                                                                                                                                                           \
+    if (verbose) {                                                                                                                                             \
+      bpf_printk("pyperf: " fmt, ##__VA_ARGS__);                                                                                                               \
+    }                                                                                                                                                          \
+  })
+
+static __always_inline long unsigned int read_tls_base(struct task_struct *task) {
+  long unsigned int tls_base;
+// This changes depending on arch and kernel version.
+// task->thread.fs, task->thread.uw.tp_value, etc.
+#if __TARGET_ARCH_x86
+  tls_base = BPF_CORE_READ(task, thread.fsbase);
+#elif __TARGET_ARCH_arm64
+  tls_base = BPF_CORE_READ(task, thread.uw.tp_value);
+#else
+#error "Unsupported platform"
+#endif
+  return tls_base;
+}
+
+//
+//   ╔═════════════════════════════════════════════════════════════════════════╗
+//   ║ BPF Programs                                                            ║
+//   ╚═════════════════════════════════════════════════════════════════════════╝
+//
+SEC("perf_event")
+int unwind_python_stack(struct bpf_perf_event_data *ctx) {
+  u64 zero = 0;
+  unwind_state_t *unwind_state = bpf_map_lookup_elem(&heap, &zero);
+  if (unwind_state == NULL) {
+    LOG("[error] unwind_state is NULL, should not happen");
+    return 1;
+  }
+
+  // @norelease: DRY
+  u64 pid_tgid = bpf_get_current_pid_tgid();
+  pid_t pid = pid_tgid >> 32;
+  pid_t tid = pid_tgid;
+  if (pid == 0) {
+    return 0;
+  }
+
+  InterpreterInfo *interpreter_info = bpf_map_lookup_elem(&pid_to_interpreter_info, &pid);
+  if (!interpreter_info) {
+    LOG("[error] interpreter_info is NULL, not a Python process or unknown Python version");
+    return 0;
+  }
+
+  if (interpreter_info->thread_state_addr == 0) {
+    LOG("[error] interpreter_info.thread_state_addr was NULL");
+    return 0;
+  }
+
+  LOG("[start]");
+  LOG("[event] pid=%d tid=%d", pid, tid);
+
+  GET_STATE();
+
+  // Reset state.
+  state->interpreter_info = (InterpreterInfo){0};
+  state->interpreter_info = *interpreter_info;
+
+  state->thread_state = 0;
+  state->current_pthread = 0;
+
+  // state->base_stack = base_stack;
+  state->frame_ptr = 0;
+  state->stack_walker_prog_call_count = 0;
+
+  // state->sample = (Sample){0};
+  state->sample.timestamp = bpf_ktime_get_ns();
+  state->sample.tid = tid;
+  state->sample.pid = pid;
+  state->sample.cpu = bpf_get_smp_processor_id();
+  state->sample.stack_status = STACK_COMPLETE;
+
+  state->sample.stack = (stack_trace_t){0};
```

Comment:
nit: 
```suggestion
  state->sample.stack = {0};
```

should work, but not sure again about the perf implications as it might introduce a mem

Reply:
```
cpu/pyperf.bpf.c:152:25: error: expected expression
  state->sample.stack = {0};
  ```